Libplanet/Consensus/ValidatorSet.cs

```diff
@@ -110,18 +110,84 @@ public BigInteger GetValidatorsPower(List<PublicKey> publicKeys)
         public bool Contains(Validator validator) => Validators.Contains(validator);
 
+        /// <summary>
+        /// Checks if <see cref="Validator"/> of given <paramref name="publicKey"/>
+        /// is a member of <see cref="ValidatorSet"/>.
+        /// </summary>
+        /// <param name="publicKey">The <see cref="PublicKey"/> to check.</param>
+        /// <returns><see langword="true"/>
+        /// if <see cref="Validator"/> of given <paramref name="publicKey"/>
+        /// is in <see cref="Validators"/>, <see langword="false"/> otherwise.</returns>
+        public bool PublicKeyContains(PublicKey publicKey)
+            => Validators.Exists(v => v.PublicKeyEquals(publicKey));
+
+        /// <summary>
+        /// Checks if given <paramref name="validator"/> is a member of <see cref="ValidatorSet"/>.
+        /// Checks only <see cref="Validator.PublicKey"/>.
+        /// </summary>
+        /// <param name="validator">The <see cref="Validator"/> to check.</param>
+        /// <returns><see langword="true"/> if given <paramref name="validator"/>
+        /// is in <see cref="Validators"/>, <see langword="false"/> otherwise.</returns>
+        public bool PublicKeyContains(Validator validator)
+            => Validators.Exists(v => v.PublicKeyEquals(validator));
+
+        /// <summary>
+        /// Checks if given <paramref name="other"/> is same as this <see cref="ValidatorSet"/>.
+        /// Checks only <see cref="Validator.PublicKey"/>.
+        /// </summary>
+        /// <param name="other">The <see cref="ValidatorSet"/> to check.</param>
+        /// <returns><see langword="true"/>
+        /// if <see cref="PublicKey"/>s of given <paramref name="other"/>
+        /// is equals to <see cref="Validators"/>, <see langword="false"/> otherwise.</returns>
+        public bool PublicKeyEquals(ValidatorSet? other) =>
+            other is ValidatorSet validators && PublicKeys.SequenceEqual(validators.PublicKeys);
+
+        /// <summary>
+        /// Create new <see cref="ValidatorSet"/> that given validator has been added.
+        /// Original <see cref="ValidatorSet"/> does not change.
+        /// </summary>
+        /// <param name="validator">The validator to be added.</param>
+        /// <returns>New <see cref="ValidatorSet"/> that given validator has been added.</returns>
+        public ValidatorSet Add(Validator validator)
+            => new ValidatorSet(Validators.Add(validator).ToList());
+
+        /// <summary>
+        /// Create new <see cref="ValidatorSet"/> that given validator has been removed.
+        /// Original <see cref="ValidatorSet"/> does not change.
+        /// On removal, check given validator's publickey and power.
+        /// </summary>
+        /// <param name="validator">The validator to be removed.</param>
+        /// <returns>New <see cref="ValidatorSet"/> that given validator has been removed.</returns>
+        public ValidatorSet Remove(Validator validator)
```

Comment:
Method does not changes instance, can you add `[Pure]` attribute to the method?

Reply:
Resolved with commit amend